### PR TITLE
Return `null` rather than `this` from INamedTypeSymbol.TupleUnderlyingType

### LINKED ISSUE #41702: INamedTypeSymbol.TupleUnderlyingType should return `null` rather than `this`
The behavior of `INamedTypeSymbol.TupleUnderlyingType` should return `null` rather than `this` when the type is a tuple without names, to avoid breaking existing callers.

If the caller recurses on `TupleUnderlyingType`, returning `this` could lead to a `StackOverflowException` that will crash the process.

The behavior changed in https://github.com/dotnet/roslyn/pull/39370.

### PR DESCRIPTION
**Pull request by cston**
Fixes #41702

### CODE REVIEW

File: src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs [hunk 1]
```diff
@@ -167,7 +167,7 @@ public interface INamedTypeSymbol : ITypeSymbol
         bool MightContainExtensionMethods { get; }
 
         /// <summary>
-        /// If this is a tuple type symbol, returns the symbol for its underlying type (ie. without element names).
+        /// If this is a tuple type with element names, returns the symbol for the tuple type without names.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs]: >If this is a tuple type with element names, returns the symbol for the tuple type without names. [](start = 12, length = 96)

Is this accurate for long tuples?

**Comment by AlekseyTs** [comment]
It looks like CI isn't passing.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/Core/Portable/Symbols/INamedTypeSymbol.cs]: Yes, that should be correct for long tuples as well.

---
In reply to: [382318717](https://github.com/dotnet/roslyn/pull/41828#discussion_r382318717) [](ancestors = 382318717)

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 2]
```diff
@@ -492,6 +491,11 @@ bool containsModopt(INamedTypeSymbol symbol)
             }
         }
 
+        private static INamedTypeSymbol GetTupleUnderlyingTypeOrSelf(INamedTypeSymbol type)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: >private static INamedTypeSymbol GetTupleUnderlyingTypeOrSelf(INamedTypeSymbol type) [](start = 8, length = 83)

It feels like we should be able to share this helper between compilers by adding it to ITypeSymbolHelpers type.

**Pull request merged**